modules/geoip/geoip-parser.c

```diff
@@ -156,20 +156,29 @@ geoip_parser_free(LogPipe *s)
   g_free(self->database);
   g_free(self->prefix);
 
-  GeoIP_delete(self->gi);
-
   log_parser_free_method(s);
 }
 
 static gboolean
 geoip_parser_init(LogPipe *s)
 {
   GeoIPParser *self = (GeoIPParser *) s;
+  GlobalConfig *cfg = log_pipe_get_config(s);
 
   geoip_parser_reset_fields(self);
 
   if (self->database)
-    self->gi = GeoIP_open(self->database, GEOIP_MMAP_CACHE);
+    {
+      GeoIP *_gi = GeoIP_open(self->database, GEOIP_MMAP_CACHE);
+      if (_gi)
+        {
+          if (self->gi)
+            GeoIP_delete(self->gi);
+          self->gi = _gi;
+        }
+      else
+        self->gi = cfg_persist_config_fetch(cfg, "geoip-obj");
```

Comment:
Offline code review has brought an issue to check before merge. Above we open the .dat file with GEOIP_MMAP_CACHE. The question is if it is still safe to to access the object when we delete the backed file in case of mmap allocation.